docs/proposals/23-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,283 @@
+requests but not limits. The HPA is driven by the average rate of API requests per pod of the target kube-apiserver
+deployment.
+
+A new component named gardener-custom-metrics is deployed on each seed. It directly scrapes metrics data from all
+ShootKapi pods on the seed and derives custom metrics based on it. The component is registered as an extension API
+service to the seed kube-apiserver. It occupies the custom metrics API extension point and is responsible for providing
+all custom metrics for the seed kube-apiserver, including the one driving the ShootKapi HPA.
+
+![Design outline](./assets/gep23-design-outline.png "Design outline")
+
+_Fig 2: Proposed design_
+
+### Element: Gardener Custom Metrics Provider Component 
+A new component, named gardener-custom-metrics is added to seed clusters. It periodically scrapes the metrics endpoints
+of all ShootKapi pods on the seed. It implements the K8s custom metrics API and provides K8s metrics specific to
+Gardener, based on custom calculations. The proposed design can naturally be extended to multiple sources of
+input data and a non-volatile cache for acquired data. However, in the scope of this proposal, the only data source
+is the aforementioned metrics scrape, and the calculated values only need to be briefly cached in memory.
+
+A kube-apiserver's metrics response measures megabytes in size. To reduce the amount of network traffic, this proposal
+utilizes compressed HTTP responses when scraping metrics. Outside the scope of this proposal, a future enhancement is
+possible, in which a metrics filter component is added as a sidecar to each ShootKapi pod. Each such sidecar would
+mirror the metrics endpoint of its respective ShootKapi, with the only addition of support for filtering based on an
+HTTP request parameter (e.g. `GET /metrics?name=apiserver_request_total` would forward a `GET /metrics` request to the
+ShootKapi, and then would reduce the response to only `apiserver_request_total` counters, before passing it on to the
+caller). 
+
+#### High availability operation
+A single gardener-custom-metrics replica does not satisfy Gardener's high availability (HA) requirements. Autoscaling
+operations can tolerate brief (1-2 minutes) periods of metrics signal outage without disruption to autoscaling service
+availability. However, node provisioning delays easily exceed that tolerable amount, and do not allow timely fail over
+via creation of a new replica after the existing one fails.
+
+For a seed in high availability mode, gardener-custom-metrics is deployed in a multi-replica active/passive arrangement,
+based on the well established controller leader election mechanism used by other Gardener components.
+Passive replicas do not respond to metrics requests. Readiness probes report "ready" status on the active replica,
+and "not ready" on passive ones. All metrics requests are routed to the one active replica, by means of a K8s service
+acting on the readiness probes.
```

Comment:
I don't think we should make readiness probes go false. This will block rollouts of the deployment and will not match the general concept behind health checks and so on.

Instead, we could relay requests to the passive instance to the active one.

Reply:
Thanks for spotting the problem at design time!